pkg/graveler/delete_sensor.go

```diff
@@ -0,0 +1,153 @@
+package graveler
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+const (
+	callbackChannelSize = 1000
+	graceDuration       = 30 * time.Second
+)
+
+type DeleteSensorCB func(repositoryID RepositoryID, branchID BranchID, stagingTokenID StagingToken, inGrace bool)
+
+type StagingTokenCounter struct {
+	StagingTokenID StagingToken
+	Counter        int
+}
+
+type stagingTokenData struct {
+	repositoryID   RepositoryID
+	branchID       BranchID
+	stagingTokenID StagingToken
+}
+
+type DeleteSensor struct {
+	cb        DeleteSensorCB
+	triggerAt int
+	callbacks chan stagingTokenData
+	wg        sync.WaitGroup
+	mutex     *sync.Mutex
+	// stopped used as flag that the sensor has stopped. stop processing CountDelete.
+	stopped                int32
+	graceDuration          time.Duration
+	branchTombstoneCounter map[RepositoryID]map[BranchID]*StagingTokenCounter
+}
+
+type DeleteSensorOpts func(s *DeleteSensor)
+
+func WithCBBufferSize(bufferSize int) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.callbacks = make(chan stagingTokenData, bufferSize)
+	}
+}
+
+func WithGraceDuration(d time.Duration) DeleteSensorOpts {
+	return func(s *DeleteSensor) {
+		s.graceDuration = d
+	}
+}
+func NewDeleteSensor(triggerAt int, cb DeleteSensorCB, opts ...DeleteSensorOpts) *DeleteSensor {
+	ds := &DeleteSensor{
+		cb:                     cb,
+		triggerAt:              triggerAt,
+		stopped:                0,
+		graceDuration:          graceDuration,
+		mutex:                  &sync.Mutex{},
+		branchTombstoneCounter: make(map[RepositoryID]map[BranchID]*StagingTokenCounter),
+		callbacks:              make(chan stagingTokenData, callbackChannelSize),
+	}
+	for _, opt := range opts {
+		opt(ds)
+	}
+	ds.wg.Add(1)
+	go ds.processCallbacks()
+	return ds
+}
+
+func (s *DeleteSensor) isStopped() bool {
+	return atomic.LoadInt32(&s.stopped) == 1
+}
+
+func (s *DeleteSensor) triggerTombstone(ctx context.Context, st stagingTokenData) {
```

Comment:
I think this is a potential major bottleneck.
The call to CountDelete is synchronous and using a lock for triggerTombstone could cause lowered performance on delete load.
I would expect the request to go into some channel which will handle the triggerTombstone asynchronously (or at least this block).

Reply:
How can that happen? it only updates a structure, there are some cases it will update a channel, but even in that case it will not wait.

Reply:
Thought about using channels at first, the two downsides were
1. If a channel is full, I need to throw the trigger away (or block, which is unacceptable)
2. It's slower with channels

Reply:
I fear @N-o-Z is right, which is both expected and unfortunate.  I dug into AWS smithy and clients are _not_ serialized per host, you could get multiple clients to the same DynamoDB server.  So without this code deletes are _not_ serialized.

However, this is actually not a particularly expensive piece of code, _especially_ in all the cases where it does not enqueue an event.  I suggest that we remain aware of it, profile it under load, and if necessary replace this map with a sync.Map.  That has a method [LoadOrStore](https://pkg.go.dev/sync#Map.LoadOrStore), which would probably do a lot of what we need here.  At worst we would need to get a bit sloppy with the count :-/

So I would like us to agree to open an issue to profile this and replace with sync.Map.